library/core/src/ptr/metadata.rs

```diff
@@ -59,6 +59,30 @@ pub trait Pointee {
     type Metadata: Copy + Send + Sync + Ord + Hash + Unpin;
 }
 
+/// Just the metadata associated with some pointee.
+///
+/// This reattaches the pointee type to the metadata to allow pointee unsizing
+/// to happen.
+///
+/// # Example
+///
+/// ```rust
+/// #![feature(ptr_metadata)]
+///
+/// use std::ptr::JustMetadata;
+///
+/// // The metadata for any sized type is ()
+/// let sized_metadata: JustMetadata<[u8; 5]> = JustMetadata::<[u8; 5]>(());
+///
+/// // But we can coerce the pointee from an array to a slice
+/// let unsized_metadata: JustMetadata<[u8]> = sized_metadata;
+///
+/// // Now the metadata is the slice length
+/// assert_eq!(unsized_metadata.0, 5);
+/// ```
+#[cfg_attr(not(bootstrap), lang = "just_metadata")]
+pub struct JustMetadata<T: ?Sized>(pub <T as Pointee>::Metadata);
```

Comment:
I am not T-libs, but perhaps this could use some helper methods like `from_raw_pointer(self, *const ()) -> *const T` which does `ptr::from_raw_parts`. I don't know though.

Also, should this derive `Copy` and other common traits?

Reply:
Deriving would use the wrong bounds (on `T` instead of `<T as Pointee>::Metadata`), but yes, this should probably get the other common traits. ~~Mostly I just wanted to open the PR as soon as I saw it working.~~

Reply:
no worries, it doesn't need to be done in this PR.